pkg/isb/forward/forward.go

```diff
@@ -285,24 +287,44 @@ func (isdf *InterStepDataForward) forwardAChunk(ctx context.Context) {
 }
 
 // ackFromBuffer acknowledges an array of offsets back to fromBuffer and is a blocking call or until shutdown has been initiated.
-func (isdf *InterStepDataForward) ackFromBuffer(ctx context.Context, offsets []isb.Offset) (err error) {
-	for {
+func (isdf *InterStepDataForward) ackFromBuffer(ctx context.Context, offsets []isb.Offset) error {
+
+	var ackRetryBackOff = wait.Backoff{
+		Factor:   1,
+		Jitter:   0.1,
+		Steps:    math.MaxInt64,
+		Duration: time.Millisecond * 10,
+	}
+	ctxClosedErr := wait.ExponentialBackoffWithContext(ctx, ackRetryBackOff, func() (done bool, err error) {
 		errs := isdf.fromBuffer.Ack(ctx, offsets)
 		summarizedErr := errorArrayToMap(errs)
+		var failedOffsets []isb.Offset
+
 		if len(summarizedErr) > 0 {
 			isdf.opts.logger.Errorw("failed to ack from buffer", zap.Any("errors", summarizedErr))
-			// TODO: implement retry with backoff etc.
-			time.Sleep(isdf.opts.retryInterval)
+
+			// retry only the failed offsets
+			for i, offset := range offsets {
+				if errs[i] != nil {
+					failedOffsets = append(failedOffsets, offset)
+				}
+			}
+			offsets = failedOffsets
```

Comment:
usually you don't do this approach, because it is mutating the function argument. It is not wrong but usually frowned upon 😄 

does it make sense to do
```go
func (isdf *InterStepDataForward) ackFromBuffer(ctx context.Context, offsets []isb.Offset) error {
   var offsets := offsets
```
to give a feeling that this is a local copy?

Reply:
@whynowy, any comments?

Reply:
does it so happen with Jetstream that if Ack operation on one offset fails but the subsequent offset may suceed concurrently/immediately?
also if it need not be in a sequence, can we not parallelize this Ack operation?

Reply:
ofcourse if the Ack is inherently async, then parallelism doesn't matter

Reply:
Ack is done concurrently https://github.com/numaproj/numaflow/blob/main/pkg/isb/stores/jetstream/reader.go#L235-L246

Reply:
The values of Slice parameter go with pointers, needs to be very careful, as long as there's no `slice[index] = xxx`, it should be okay. Line 312 is already to define a local variable, should be safe. But it would be better to avoid ambiguity :)